---
title: Structural Variation Report
author: UPD-UCSC
urlcolor: teal
output:
  html_document:
    theme: readable
---

<style type="text/css">
.main-container {
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r include=FALSE}
knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE, fig.width=10)
```

```{r setup}
## Read input arguments
## 1: config file in YAML
## 2: filename for the output TSV (annotated SVs)
args = commandArgs(TRUE)

## Load packages
library(sveval)
library(GenomicRanges)
library(knitr)
library(DT)
library(dplyr)
library(ggplot2)

## Reciproal overlap
rOverlap <- function(qgr, sgr){
  findOverlaps(qgr, sgr) %>% as.data.frame %>%
    mutate(qw=width(qgr)[queryHits], sw=width(sgr)[subjectHits],
           qsw=width(pintersect(qgr[queryHits], sgr[subjectHits])),
           rol=ifelse(qw>sw, qsw/qw, qsw/sw))
}

## Winsorize function
winsor <- function(x, u){
  if(any(x>u)) x[x>u] = u
  x
}

## Format links for genes (NCBI), genomic position (UCSC Genome Browser), or pli (gnomAD)
formatLink <- function(labels, urls, format='pdf'){
  if(format=='html'){
    return(paste0('[', labels, '](', urls,')'))
  }
  if(format=='rawhtml'){
    return(paste0('<a href="', urls, '" target="_blank">', labels, '</a>'))
  }
  if(format=='pdf'){
    return(paste0('{\\href{', urls, '}{', labels, '}}'))
  }
}
linkGenes <- function(genes, format='html'){
  urls = paste0('https://www.ncbi.nlm.nih.gov/gene?term=(', genes, '%5BGene%20Name%5D)%20AND%20Human%5BOrganism%5D')
  return(formatLink(genes, urls, format=format))
}
linkPli <- function(scores, genes, digits=3, format='html'){
  urls = paste0('https://gnomad.broadinstitute.org/gene/', genes)
  labels = round(scores, digits)
  return(formatLink(labels, urls, format=format))
}
linkPos <- function(chr, pos, type, size, flanks=500, format='html'){
  size = ifelse(type %in% c('DEL','DUP','INV', 'LOH'), size, 1)
  urls = paste0('https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=', chr, '%3A',
                pos-flanks, '-', pos+size+flanks, '&highlight=hg38.', chr, '%3A',
                pos, '-', pos+size, '%23E18F99')
  labels = paste0(chr, ':', pos)
  return(formatLink(labels, urls, format=format))
}

## Format some fields of a data.frame to convert them into links to external resources
formatTable <- function(df){
  if(nrow(df)==0) return(df)
  ## make some columns "factors" to help filtering
  for(coln in c('type', 'ac', 'method', 'dgv', 'meths')){
    if(coln %in% colnames(df)){
      df[,coln] = factor(df[,coln])
    }
  }
  if('pos' %in% colnames(df)){
    df$pos = linkPos(df$chr, df$pos, df$type, df$size, format='rawhtml')
    df$chr = NULL
  }
  if('gene' %in% colnames(df)){
    if('pLI' %in% colnames(df)){
      df$pLI = linkPli(df$pLI, df$gene, format='rawhtml')
    }
    df$gene = linkGenes(df$gene, format='rawhtml')
  }
  df
}

## Read config file
cfg.r = scan(args[1], '', sep='\n', quiet=TRUE)
cfg.r = strsplit(cfg.r, ': ')
cfg = gsub('"', '', unlist(lapply(cfg.r, '[', 2)))
names(cfg) = unlist(lapply(cfg.r, '[', 1))

## for debug:
## cfg = c(sniffles_vcf='na12878.sniffles.vcf', svim_vcf='na12878.svim.vcf',
##         idxcov='indexcov_na12878-indexcov.bed.gz', gene_pos='gene_position_info.tsv',
##         pli_gene='gnomad.v2.1.1.lof_metrics.by_gene.txt.bgz',
##         simprep='simpleRepeat.txt.gz',
##         hsvlr='hsvlr.vcf.gz',
##         dgv='GRCh38_hg38_variants_2016-08-31.txt',
##         clingen='iscaPathogenic.txt.gz',
##         ctcf='ENCFF010WHH.bed.gz',
##         gnomadsv='gnomad_v2_sv.sites.pass.lifted.vcf.gz',
##         cre='ENCFF166QIT.lifted.bed.gz',
##         cons='phastConsElements100way.txt.gz',
##         gencode='gencode.v35.annotation.gtf.gz',
##         cytoband='cytoBandIdeo.txt.gz')

message('Read VCFs...')
svs.gr.l = list()
if('sniffles_vcf' %in% names(cfg)){
  svs.gr.l$sniffles = readSVvcf(cfg['sniffles_vcf'])
  svs.gr.l$sniffles$method = 'Sniffles'
}
if('svim_vcf' %in% names(cfg)){
  svs.gr.l$svim = readSVvcf(cfg['svim_vcf'])
  svs.gr.l$svim$method = 'SVIM'
}
if('cnv_ill' %in% names(cfg)){
  svs.gr.l$freec = readSVvcf(cfg['cnv_ill'])
  svs.gr.l$freec$ac = 1
  svs.gr.l$freec$method = 'Control-FREEC'
}
if('smoove_ill' %in% names(cfg)){
  svs.gr.l$smoove = readSVvcf(cfg['smoove_ill'])
  svs.gr.l$smoove$method = 'Smoove'
}

message('Combine variants...')
svs.gr = Reduce(c, svs.gr.l)
svs.gr = subset(svs.gr, ac>0)

message('Annotate overlap between methods...')
ol.gr = svOverlap(svs.gr, svs.gr, min.ol=.5, max.ins.dist=100)
ol.df = ol.gr %>% as.data.frame %>%
  mutate(method=svs.gr$method[subjectHits]) %>%
  group_by(queryHits) %>%
  summarize(nb.meths=length(unique(method)),
            meths=paste(sort(unique(method)), collapse=' '))
svs.gr$nb.meths = 1
svs.gr$nb.meths[ol.df$queryHits] = ol.df$nb.meths
svs.gr$meths = svs.gr$method
svs.gr$meths[ol.df$queryHits] = ol.df$meths

message('Annotate frequency using gnomAD...')
gnomad = readSVvcf(cfg['gnomadsv'], other.field='AF')
gnomad$ac = 1
ol.df = svOverlap(svs.gr, gnomad, min.ol=.1, max.ins.dist=100) %>% as.data.frame %>%
  mutate(freq=gnomad$AF[subjectHits]) %>%
  group_by(queryHits) %>%
  summarize(freq=max(freq))
svs.gr$freq = 0
svs.gr$freq[ol.df$queryHits] = ol.df$freq

message('Gene annotation...')
types.ranked = c('CDS', 'UTR', 'promoter', 'gene')
types.labels = c('coding', 'UTR', 'promoter', 'intronic')
genc = rtracklayer::import(cfg['gencode'])
genc = subset(genc, type %in% types.ranked)
prom = promoters(subset(genc, type=='gene'))
prom$type = 'promoter'
genc = c(genc, prom)

ol.genc = findOverlaps(svs.gr, genc) %>% as.data.frame %>%
  mutate(gene=genc$gene_name[subjectHits],
         gene_type=genc$gene_type[subjectHits],
         type=genc$type[subjectHits],
         type=factor(type, levels=types.ranked, labels=types.labels)) %>%
  arrange(type) %>% 
  group_by(queryHits, gene, gene_type) %>%
  summarize(impact=head(type, 1))

ol.genc.sum = ol.genc %>% group_by(queryHits) %>%
  summarize(genes=ifelse(n()>10,
                         paste(n(), 'genes'),
                         paste0(gene, '(', impact, '>', gene_type, ')', collapse=' ')),
            nb.pc.genes=sum(gene_type=='protein_coding'))
svs.gr$genes = NA
svs.gr$genes[ol.genc.sum$queryHits] = ol.genc.sum$genes
svs.gr$nb.pc.genes = 0
svs.gr$nb.pc.genes[ol.genc.sum$queryHits] = ol.genc.sum$nb.pc.genes

## Nearby protein-coding genes
pc.genes = subset(genc, type=='gene' & gene_type=='protein_coding')
d.df = distanceToNearest(svs.gr, pc.genes, ignore.strand=TRUE) %>% as.data.frame
svs.gr$gene.dist = NA
svs.gr$gene.dist[d.df$queryHits] = d.df$distance
svs.gr$gene.near = NA
svs.gr$gene.near[d.df$queryHits] = pc.genes$gene_name[d.df$subjectHits]

## Near coding from protein-coding gene
pc.cds = subset(genc, type=='CDS' & gene_type=='protein_coding')
d.df = distanceToNearest(svs.gr, pc.cds, ignore.strand=TRUE) %>% as.data.frame
svs.gr$cds.dist = NA
svs.gr$cds.dist[d.df$queryHits] = d.df$distance

message('Import genes of interest...')
genes = read.table(cfg['gene_pos'], as.is=TRUE, header=TRUE) %>%
  dplyr::rename(gene=gene_name) %>%
  mutate(gene_list=factor(gene_list))
genes.list = genes %>% select(gene, gene_list) %>% unique
genes.gr = makeGRangesFromDataFrame(genes, keep.extra.columns=TRUE)

d.df = distanceToNearest(svs.gr, genes.gr, ignore.strand=TRUE) %>% as.data.frame
svs.gr$sel.gene.dist = NA
svs.gr$sel.gene.dist[d.df$queryHits] = d.df$distance
svs.gr$sel.gene.near = NA
svs.gr$sel.gene.near[d.df$queryHits] = genes.gr$gene[d.df$subjectHits]

message('Import pli scores...')
pli.df = read.table(cfg['pli_gene'], as.is=TRUE, header=TRUE, sep='\t')
pli.df = pli.df %>% select(gene, pLI) %>% unique

message('Overlap with simple repeats...')
sr = read.table(cfg['simprep'])
sr = sr[,c(2,3,4,6,7,17)]
colnames(sr) = c('chrom', 'start', 'end', 'period', 'copyNum', 'sequence')
sr = makeGRangesFromDataFrame(sr, keep.extra.columns=TRUE)
svs.gr$simp.rep = overlapsAny(svs.gr, sr, maxgap=10)

message('Overlap with SVs from public long-read studies...')
lr.gr = readSVvcf(cfg['hsvlr'])
lr.gr$ac = 1
ol.gr = svOverlap(svs.gr, lr.gr, min.ol=.5, max.ins.dist=100)
svs.gr$hsvlr = FALSE
svs.gr$hsvlr[ol.gr$queryHits] = TRUE

message('Overlap with DGV catalog...')
dgv = read.table(cfg['dgv'], as.is=TRUE, header=TRUE, sep='\t')
dgv = dgv[,c('chr','start','end', 'variantsubtype')]
dgv$chr = paste0('chr', dgv$chr)
dgv = makeGRangesFromDataFrame(dgv, keep.extra.columns=TRUE)
dgv.loss = subset(dgv, variantsubtype %in% c('loss', 'deletion', 'gain+loss'))
dgv.loss = overlapsAny(svs.gr, dgv.loss, maxgap=10)
dgv.gain = subset(dgv, variantsubtype %in% c('duplication', 'gain', 'insertion',
                                             'mobile element insertion',
                                             'novel sequence insertion',
                                             'tandem duplication', 'gain+loss'))
dgv.gain = overlapsAny(svs.gr, dgv.gain, maxgap=10)
svs.gr$dgv = 'none'
svs.gr$dgv = ifelse(dgv.loss, 'loss', svs.gr$dgv)
svs.gr$dgv = ifelse(dgv.gain, 'gain', svs.gr$dgv)
svs.gr$dgv = ifelse(dgv.gain & dgv.loss, 'loss+gain', svs.gr$dgv)

message('Overlap with ClinGenn pathogenic CNVs...')
clingen = read.table(cfg['clingen'], as.is=TRUE, sep='\t')
clingen = GRanges(clingen$V2, IRanges(clingen$V3, clingen$V4))
ol.o = rOverlap(svs.gr, clingen)
svs.gr$clingen = FALSE
svs.gr$clingen[subset(ol.o, rol>.5)$queryHits] = TRUE

message('Overlap with CTCF peaks...')
ctcf = read.table(cfg['ctcf'], as.is=TRUE)
ctcf = with(ctcf, GRanges(V1, IRanges(V2, V3), score=V5))
svs.gr$ctcf = overlapsAny(svs.gr, ctcf)

message('Overlap with regulatory regions...')
cres = read.table(cfg['cre'], as.is=TRUE)
cres = with(cres, GRanges(V1, IRanges(V2, V3), score=V5))
svs.gr$cres = overlapsAny(svs.gr, cres)

message('Overlap with conserved regions...')
cons.gr = read.table(cfg['cons'], as.is=TRUE)
cons.gr = with(cons.gr, GRanges(V2, IRanges(V3, V4)))
svs.gr$cons = overlapsAny(svs.gr, cons.gr)

## clean up some memory
rm(cons.gr, cres, ctcf, clingen, dgv, lr.gr, sr, gnomad)
dump = capture.output(gc())

svs.gr$cov = NA
if('idxcov' %in% names(cfg)){
  message('Read coverage stats from indexcov...')
  cov.df = read.table(cfg['idxcov'], as.is=TRUE, header=TRUE, sep='\t', comment='')
  colnames(cov.df) = c('chr', 'start', 'end', 'cov')
  
  chrs.order = c(1:22, 'X','Y','M')
  if(all(grepl('chr', unique(cov.df$chr)))){
    chrs.order = paste0('chr', chrs.order)
  }
  cov.df = cov.df %>% mutate(chr=factor(chr, levels=chrs.order))

  nb.bins = 10
  cov.10.df = cov.df %>%
    filter(chr!='chrM') %>% 
    arrange(chr, start) %>%
    group_by(chr) %>% mutate(bin=head(rep(1:(n()/nb.bins+1), each=nb.bins), n())) %>%
    group_by(chr, bin) %>% summarize(start=min(start), end=max(end), cov=sum(cov)) %>%
    ungroup %>% mutate(cov=cov/median(cov))

  cov.gr = makeGRangesFromDataFrame(cov.10.df, keep.extra.columns=TRUE)
  idx.large = which(svs.gr$size>1e5)
  ol.df = findOverlaps(svs.gr[idx.large], cov.gr) %>% as.data.frame %>%
    mutate(cov=cov.gr$cov[subjectHits]) %>%
    group_by(queryHits) %>% summarize(cov=paste0(round(median(cov), 2), ' (', n(), ')'))
  svs.gr$cov[idx.large[ol.df$queryHits]] = ol.df$cov
}

## Gene-level info
svs.gene = svs.gr[ol.genc$queryHits]
svs.gene$genes = NULL
svs.gene$gene = ol.genc$gene
svs.gene$gene_type = ol.genc$gene_type
svs.gene$impact = ol.genc$impact
```

# {.tabset}

## Methods

#### SV calling 

The structural variants were called from the nanopore reads using [Sniffles](https://github.com/fritzsedlazeck/Sniffles) and [SVIM](https://github.com/eldariont/svim).
[indexcov](https://github.com/brentp/goleft/tree/master/indexcov) provides quick estimates of read coverage to confirm large CNVs or identify chromosomal aberrations. 
If short-read data is available, we run [Smoove](https://github.com/brentp/smoove) (~Lumpy) and [Control-FREEC](http://boevalab.inf.ethz.ch/FREEC/) (CNV and LOH detection).
To focus on high-confidence SV calls, most tables only look at calls made by at least two methods

#### SV database and frequency annotation

The variants were compared to catalogs of known SVs.
The frequency estimates are based on the [gnoma-SV catalog](https://macarthurlab.org/2019/03/20/structural-variants-in-gnomad/), as the maximum frequency of variants with reciprocal overlap >10%.
A few samples have been sequenced with long-read sequencing techonology (e.g. PacBio or ONT): HGSVC: [Chaisson et al 2019](https://pubmed.ncbi.nlm.nih.gov/30992455/), SVPOP: [Audano et al. 2019](https://pubmed.ncbi.nlm.nih.gov/30661756/), GIAB: [Zook et al. 2020](https://pubmed.ncbi.nlm.nih.gov/32541955/).
We filter out calls that match these public catalogs to remove variants that pass the frequency filter based on gnomAD-SV simply because they can't be detected by short reads.
Finally, variants are flagged if overlapping [DGV](http://dgv.tcag.ca/dgv/app/home) (any overlap) or [ClinGenn pathogenic variants](https://www.clinicalgenome.org/) (pathogenic set, reciprocal overlap>50%).

#### Gene annotation

The [GENCODE](https://www.gencodegenes.org/) gene annotation was used to flag variants as *coding*, *UTR*, *promoter*, *intronic* (prioritized in this order) and compute the distance to the nearest gene.
While we consider lncRNA, miRNA in the annotation (`genes` column), most tables focus on protein-coding genes.

The *pLI* score was computed by the [gnomAD project](https://gnomad.broadinstitute.org/).
It represents the probability that the gene is intolerant to loss-of-function variants.
A variant affecting a gene with a high pLI score (e.g. >0.9) is more likely to have a biological impact.
The variants in each section are ordered to show those affecting genes with the highest pLI first.

#### Filters

In most tables, we removed common variants, i.e. either:

- frequency higher than 1% in gnomAD-SV
- seen in the SV catalog from long-read studies

#### Column names

- `pLI` prob of  loss-of-function intolerance described above (`-1` if no information available).
- `type` SV type
- `ac` allele count: `1` for het or `2` for hom.
- `freq` allele frequency of similar SV in gnomAD-SV (>10K genomes sequenced with Illumina whole-genome sequencing).
- `dgv` does the variant overlap any variant in DGV in any way?
- `clingen` any similar ClinGen pathogenic variants (as found at the UCSC Genome Browser)? "Similar" defined as reciprocal overlap > 50%.
- `ctcf` does the variant overlap a CTCF binding site? From ENCODE track for kidney.
- `cres` does the variant overlap a regulatory region? From ENCODE track for kidney.
- `simp.rep` is the variant in or close to a simple repeat (see simple repeat track in the UCSC Genome Browser).
- `hsvlr` any similar variant in SV catalogs from public long-read sequencing studies?
- `cons` does the variant overlap a conserved element (as defined by 100 vertebrate phastCons track)
- `impact` potential impact based on gene annotation: *coding*, *UTR*, *promoter*, *intronic*.
- `genes` a summary of the genes overlapped by the variant and their impact.
- `gene.dist` distance to the nearest gene which is specified by `gene.near`.
- `sel.gene.dist` distance to the nearest gene of interest which is specified by `sel.gene.near`.
- `cds.dist` distance to the nearest coding regions, useful for intronic variants for example.
- `nb.pc.genes` number of protein-coding genes overlapped by the variant.
- `method` the method that made the SV call
- `meths` all the methods that called the SV or a similar one
- `nb.meths` the number of methods that called the SV or a similar one
- `cov` median scaled coverage for large variants (>100 kbp), if *indexcov* results are available. Between parenthesis is the number of bin overlapping the variant (the higher the more confident).

#### TSV file

In addition to the report, a gene-centric TSV file is written with all the annotation described above.
Gene-centric means that there is one row for each gene-variant pair.
This helps filter on gene features (e.g. gene of interest, pLI).

## Genes

There are `r length(unique(genes$gene))` genes in `r length(unique(genes$gene_list))` gene lists:

```{r genes}
genes %>% group_by(gene_list) %>% summarize(genes=n()) %>% kable
```

---

They contain the following genes:

```{r genes2}
genes %>% formatTable %>% datatable(filter='top', escape=FALSE, options=list(pageLength=25, autoWidth=TRUE))
```

## Coding/UTR/promoter

Variants overlapping coding, UTR or promoter ([TSS-2000bp,TSS+200bp]) of **protein-coding genes**, including genes of interest.
Gene-centric table: one row per affected gene, which means a variant can contribute multiple rows in the table.

```{r table_high_gene}
svs.gene %>% as.data.frame %>% mutate(chr=seqnames, pos=start) %>%
  filter(nb.meths>1, gene_type=='protein_coding',
         freq<.01, !hsvlr, impact %in% c('coding', 'UTR', 'promoter')) %>%
  merge(pli.df, all.x=TRUE) %>%
  mutate(pLI=ifelse(is.na(pLI), -1, pLI)) %>% 
  merge(genes.list, all.x=TRUE) %>% 
  select(gene_list, gene, impact, pLI,
         pos, type, size, ac, freq,
         simp.rep, dgv, cons, cres,
         method, nb.meths, meths, chr) %>%
  arrange(is.na(gene_list), impact, desc(pLI), freq, desc(size)) %>%
  formatTable() %>%
  datatable(filter='top', escape=FALSE, options=list(pageLength=25))
```

## Intronic

Variants overlapping intronic regions but no coding, UTR or promoter of **protein-coding genes**.
We filtered out intronic variants that are not within genes of interest AND don't overlap a regularoty/conserved region AND are located further than 1 kbp from coding sequences.
Gene-centric table: one row per affected gene, which means a variant can contribute multiple rows in the table.

```{r table_intronic}
svs.gene %>% as.data.frame %>% mutate(chr=seqnames, pos=start) %>%
  filter(nb.meths>1, gene_type=='protein_coding',
         freq<.01, !hsvlr, impact == 'intronic') %>%
  merge(pli.df, all.x=TRUE) %>%
  mutate(pLI=ifelse(is.na(pLI), -1, pLI)) %>% 
  merge(genes.list, all.x=TRUE) %>%
  filter(!is.na(gene_list) | cds.dist<1e3 | cons | cres) %>% 
  select(gene_list, gene, pLI, cds.dist,
         pos, type, size, ac, freq,
         simp.rep, dgv, cons, cres,
         method, nb.meths, meths, chr) %>%
  arrange(is.na(gene_list), cds.dist, desc(pLI), freq, desc(size)) %>%
  formatTable() %>%
  datatable(filter='top', escape=FALSE, options=list(pageLength=25))
```

## Rare of interest

All variants that don't overlap genes of interest but are close (<100kbp).

```{r table_rare_near_sel}
svs.gr %>% as.data.frame %>% mutate(chr=seqnames, pos=start) %>%
  filter(nb.meths>1, sel.gene.dist>0, sel.gene.dist<1e5, freq<.01) %>%
  select(sel.gene.dist, sel.gene.near, 
         pos, type, size, ac, freq, genes,
         simp.rep, dgv, cons, cres, ctcf,
         method, nb.meths, meths, chr) %>%
  arrange(sel.gene.dist, freq) %>%
  formatTable() %>%
  datatable(filter='top', escape=FALSE, options=list(pageLength=25))
```

## Conserved/regulated

Intergenic variants overlapping conserved regions or known regulatory regions.

```{r table_noncoding}
svs.gr %>% as.data.frame %>% mutate(chr=seqnames, pos=start) %>%
  filter(nb.meths>1, freq<.01, !hsvlr, cres | cons | ctcf,
         gene.dist>0) %>%
  select(pos, type, size, ac, freq, genes,
         cons, cres, ctcf,
         gene.dist, gene.near,
         simp.rep, dgv, 
         method, nb.meths, meths, chr) %>%
  arrange(freq, gene.dist, !cons, !cres, desc(size)) %>%
  formatTable() %>%
  datatable(filter='top', escape=FALSE, options=list(pageLength=25))
```

## Large

Large and rare variants tend to have a higher biological impact.
This table shows rare variants larger than 1kbp.
Interesting profiles to look for:

- large deletions spanning a CTCF binding region (could lead to TAD reorganization and ectopic gene expression).
- Overlap with pathogenic CNV in the ClinGen database. *clingen* column shows variants with 50% reciprocal overlap with a pathogenic CNV.
- Large SV affecting multiple protein-coding genes.

```{r rarelarge}
svs.gr %>% as.data.frame %>% mutate(chr=seqnames, pos=start) %>%
  filter(size>1e3, freq<.01, !hsvlr, nb.meths>1) %>%
  select(pos, type, size, ac, freq, 
         nb.pc.genes, genes,
         clingen, cons, cres, ctcf,
         simp.rep, dgv, cov,
         method, nb.meths, meths, chr) %>%
  arrange(!clingen, desc(nb.pc.genes), desc(abs(size))) %>%
  formatTable %>%
  datatable(filter='top', escape=FALSE, options=list(pageLength=25))
```


```{r writetsv}
## remove duplicates, inc. when differring only with effect (second being NA)
svs.all = subset(svs.gr, is.na(genes))
svs.all$genes = NULL
svs.all$gene = svs.all$gene_type = svs.all$impact = NA
svs.all = c(svs.gene, svs.all)

## write TSV
out.tsv = svs.all %>% as.data.frame %>% unique
write.table(out.tsv, file=args[2], sep='\t', row.names=FALSE, quote=FALSE)
```

## Chr. aberrations

If `indexcov` was run: coverage estimates across the chromosomes to identify chromosomal aberrations 

Of note, some arms are not well assembled so it's normal to see low coverage in: 13p, 14p, 15p, 22p.
Same for the the last chunk of chrYq and the beginning og 21p.

```{r idxcov, fig.height=10}
if('idxcov' %in% names(cfg)){
  nb.bins = 50
  cov.df = cov.df %>%
    filter(chr!='chrM') %>% 
    arrange(chr, start) %>%
    group_by(chr) %>% mutate(bin=head(rep(1:(n()/nb.bins+1), each=nb.bins), n())) %>%
    group_by(chr, bin) %>% summarize(start=min(start), end=max(end), cov=sum(cov)) %>%
    ungroup %>% mutate(cov=cov/median(cov))

  cat('Average region size in the graph below is: ', round(mean(cov.df$end-cov.df$start)/1e3, 2), 'kbp')
  
  ## Arm and centromere
  cyto.df = read.table(cfg['cytoband'], as.is=TRUE, sep='\t')
  colnames(cyto.df) = c('chr', 'start', 'end', 'band', 'gieaStain')
  cyto.df = cyto.df %>% mutate(arm=substr(band, 1, 1)) %>%
    filter(arm %in% c('p', 'q')) %>% 
    group_by(chr, arm) %>% summarize(start=min(start), end=max(end))
  cyto.gr = makeGRangesFromDataFrame(cyto.df, keep.extra.columns=TRUE)

  cov.gr = makeGRangesFromDataFrame(cov.df)
  cov.df$arm = 'other'
  cov.df$arm[overlapsAny(cov.gr, subset(cyto.gr, arm=='p'))] = 'p'
  cov.df$arm[overlapsAny(cov.gr, subset(cyto.gr, arm=='q'))] = 'q'
  
  ggp = ggplot(cov.df, aes(x=start/1e6, y=winsor(cov, 2))) +
    geom_point(aes(color=arm), alpha=.5) + theme_bw() +
    geom_hline(yintercept=1, linetype=2) + 
    geom_smooth(se=FALSE, color='black') +
    facet_wrap(~chr, scales='free') +
    ylab('normalized coverage') +
    xlab('position (Mbp)') +
    ylim(0,2) +
    scale_colour_brewer(palette='Set1')
    
  print(ggp)

  cov.df %>% group_by(chr, arm) %>% summarize(median.cov=median(cov)) %>%
    arrange(desc(abs(1-median.cov))) %>%
    datatable(filter='top', escape=FALSE, options=list(pageLength=25))
}
```

## LOH

If Control-FREEC was run, it might have identified regions with loss of heterozygosity. 
Note: these calls are less confident than the rest because only deteted by one method.

```{r table_loh}
svs.gene %>% as.data.frame %>% mutate(chr=seqnames, pos=start) %>%
  filter(type=='LOH') %>%
  merge(pli.df, all.x=TRUE) %>%
  mutate(pLI=ifelse(is.na(pLI), -1, pLI)) %>% 
  merge(genes.list, all.x=TRUE) %>% 
  select(gene_list, gene, impact, pLI,
         pos, size, type,
         simp.rep, dgv, cons, cres,
         chr) %>%
  arrange(is.na(gene_list), impact, desc(pLI), desc(size)) %>%
  formatTable() %>%
  datatable(filter='top', escape=FALSE, options=list(pageLength=25))
```
