---
title: Structural Variation Report
author: UPD-UCSC
header-includes:
- \usepackage{booktabs}
- \usepackage{makecell}
urlcolor: teal
output:
  pdf_document:
    toc: true
    toc_depth: 2
---

```{r include=FALSE}
knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE, fig.width=10)
```

```{r setup}
## Read input arguments
## 1: VCF file from Sniffles
## 2: VCF file from SVIM
## 3: gene list file
## 4: gene and LoF intolerance score file
args = commandArgs(TRUE)
## args = c('na12878-chr11.sniffles.ann.freqGnomADcov10.vcf','na12878-chr11.svim.ann.freqGnomADcov10.vcf','gene_position_info.txt', 'gnomad.v2.1.1.lof_metrics.by_gene.txt.bgz')

## Load packages
library(sveval)
library(VariantAnnotation)
library(dplyr)
library(knitr)
library(kableExtra)

## Parse the SnpEff field EFF
parseEff <- function(eff){
  effect = gsub('(.*)\\(.*\\)', '\\1', eff)
  target = gsub('.*\\((.*)\\)', '\\1', eff)
  target = strsplit(target, '\\|')[[1]]
  return(tibble(effect=effect, impact=target[[1]], gene=target[[6]],
              target.type=target[[7]]))
}
## Format links for genes (NCBI) or genomic position (UCSC Genome Browser)
mdLinkGenes <- function(genes){
  paste0('[', genes, '](https://www.ncbi.nlm.nih.gov/gene?term=(', genes, '%5BGene%20Name%5D)%20AND%20Human%5BOrganism%5D)')
}
latexLinkGenes <- function(genes){
  paste0('{\\href{https://www.ncbi.nlm.nih.gov/gene?term=(', genes, '%5BGene%20Name%5D)%20AND%20Human%5BOrganism%5D}{', genes, '}}')
}
latexLinkPli <- function(scores, genes, digits=3){
  paste0('{\\href{https://gnomad.broadinstitute.org/gene/', genes, '}{', round(scores, digits), '}}')
}
mdLinkPli <- function(scores, genes, digits=3){
  paste0('[',round(scores, digits), '](https://gnomad.broadinstitute.org/gene/', genes,')')
}
latexLinkPos <- function(chr, pos){
  paste0('{\\href{https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=', chr, '%3A', pos-500, '-', pos+500, '}{', chr, ':', pos, '}}')
}
col.width='13cm'
col.width.gene='5cm'

## Read both VCFs
vcf.sn = readSVvcf(args[1], vcf.object=TRUE)
vcf.sv = readSVvcf(args[2], vcf.object=TRUE)

## Make GRanges object for the overlap
gr.sn = rowRanges(vcf.sn)
end(gr.sn) = info(vcf.sn)$END
gr.sn$size = info(vcf.sn)$SIZE
gr.sn$type = info(vcf.sn)$SVTYPE
gr.sn$GT = 'hom'
gr.sn$ID = 1:length(vcf.sn)
gr.sv = rowRanges(vcf.sv)
end(gr.sv) = info(vcf.sv)$END
gr.sv$size = info(vcf.sv)$SIZE
gr.sv$type = info(vcf.sv)$SVTYPE
gr.sv$GT = 'hom'
gr.sv$ID = 1:length(vcf.sv)

## Overlap and add a column with overlap
ol.o = svOverlap(gr.sn, gr.sv)
info(vcf.sn)$METHODS = 1
info(vcf.sn)$METHODS[ol.o$query$ID] = 2
info(vcf.sv)$METHODS = 1
info(vcf.sv)$METHODS[ol.o$subject$ID] = 2
info(vcf.sn)$METHOD = 'Sniffles'
info(vcf.sv)$METHOD = 'SVIM'

## Read gene list
genes = read.table(args[3], as.is=TRUE, header=TRUE)

## Read pli score
pli.df = read.table(args[4], as.is=TRUE, header=TRUE, sep='\t')
pli.df = pli.df %>% select(gene, pLI)

## FOR TESTING PURPOSE
genes$hgnc_symbol[1] = 'MUC5B'
```

```{r highimpact}
## Select variants with "HIGH" impact and parse the EFF field
eff.df = lapply(list(vcf.sn, vcf.sv), function(vcf){
  high.ii = which(unlist(lapply(info(vcf)$EFF, function(effs) any(grepl('HIGH', effs)))))
  eff.df = lapply(high.ii, function(ii){
    effs = lapply(info(vcf)$EFF[[ii]], parseEff)
    effs = do.call(rbind, effs)
    effs$id = ii
    effs
  })
  eff.df = do.call(rbind, eff.df)
  eff.df$method = info(vcf)$METHOD[1]
  eff.df
})
eff.df = do.call(rbind, eff.df)
```

# Methods

The structural variants found by [Sniffles](https://github.com/fritzsedlazeck/Sniffles) and [SVIM](https://github.com/eldariont/svim) were annotated by [SnpEff](http://snpeff.sourceforge.net/) and with the frequency of variants in the [gnoma-SV catalog](https://macarthurlab.org/2019/03/20/structural-variants-in-gnomad/).

The variants' "effects" in this report come from SnpEff and follow the format [described in SnpEff documentation](http://snpeff.sourceforge.net/SnpEff_manual.html#eff).
In this report we select variants with a predicted *HIGH* effect.

The *pLI* score was computed by the [gnomAD project](https://gnomad.broadinstitute.org/).
It represents the probability that the gene is intolerant to loss-of-function variants.
A variant affecting a gene with a high pLI score (e.g. >0.9) is more likely to have a biological impact.
The variants in each section are ordered to show those affecting genes with the highest pLI first.

Note: for testing purposes I added *MUC5B* to the list of genes.

# By variant

## High impact variants in genes of interest

Genes of interest: `r sort(genes$hgnc_symbol)`.

```{r genesel, results='asis'}
high.df = eff.df %>% filter(impact=='HIGH', gene %in% genes$hgnc_symbol) %>%
  merge(pli.df, all.x=TRUE) %>% arrange(desc(pLI))
  
high.ii = high.df %>% select(id, method) %>% unique

## Write a section for each variant
t = lapply(1:nrow(high.ii), function(ii){
  id.ii = high.ii$id[ii]
  if(high.ii$method[ii] == 'Sniffles'){
    vcf = vcf.sn[id.ii]
    reads = info(vcf.sn)$RE[id.ii]
  } else {
    vcf = vcf.sv[id.ii]
    reads = info(vcf.sv)$SUPPORT[id.ii]
  }
  ## Variant info
  res = tibble(
    pos=latexLinkPos(as.character(seqnames(rowRanges(vcf))), start(rowRanges(vcf))),
    type=info(vcf)$SVTYPE,
    size=info(vcf)$SVLEN,
    ## quality=rowRanges(vcf)$QUAL,
    reads=reads,
    gnomad.freq=info(vcf)$AFMAX,
    method=info(vcf)$METHOD,
    nb.methods=info(vcf)$METHODS)
  ## Effect info
  df = high.df %>% filter(id==id.ii) %>% select(gene, effect, pLI) %>% unique
  ## Write Markdown/Latex section
  cat('\n\n### ', res$type, 'in', paste(mdLinkGenes(unique(df$gene)), collapse=' '), '\n\n')
  res %>% kable(booktabs=TRUE, escape=FALSE) %>% cat
  cat('\n\n#### Effect(s) \n\n')
  df %>% mutate(effect=gsub('\\+', ' \\+ ', effect),
                effect=gsub('_', '\\\\_', effect),
                pLI=latexLinkPli(pLI, gene),
                gene=latexLinkGenes(gene)) %>%
    kable(booktabs=TRUE, escape=FALSE) %>% column_spec(2, width=col.width) %>% cat
})
```

## High impact variants affecting other protein-coding genes

```{r proteincoding, results='asis'}
high.df = eff.df %>% filter(impact=='HIGH', target.type=='protein_coding',
                            !(gene %in% genes$hgnc_symbol)) %>%
  merge(pli.df, all.x=TRUE) %>% arrange(desc(pLI))
high.ii = high.df %>% select(id, method) %>% unique

## Write a section for each variant
t = lapply(1:nrow(high.ii), function(ii){
  id.ii = high.ii$id[ii]
  if(high.ii$method[ii] == 'Sniffles'){
    vcf = vcf.sn[id.ii]
    reads = info(vcf.sn)$RE[id.ii]
  } else {
    vcf = vcf.sv[id.ii]
    reads = info(vcf.sv)$SUPPORT[id.ii]
  }
  ## Variant info
  res = tibble(
    pos=latexLinkPos(as.character(seqnames(rowRanges(vcf))), start(rowRanges(vcf))),
    type=info(vcf)$SVTYPE,
    size=info(vcf)$SVLEN,
    ## quality=rowRanges(vcf)$QUAL,
    reads=reads,
    gnomad.freq=info(vcf)$AFMAX,
    method=info(vcf)$METHOD,
    nb.methods=info(vcf)$METHODS)
  ## Effect info
  df = high.df %>% filter(id==id.ii) %>% select(gene, effect, pLI) %>% unique
  ## Write Markdown/Latex section
  cat('\n\n### ', res$type, 'in', paste(mdLinkGenes(unique(df$gene)), collapse=' '), '\n\n')
  res %>% kable(booktabs=TRUE, escape=FALSE) %>% cat
  cat('\n\n#### Effect(s) \n\n')
  df %>% mutate(effect=gsub('\\+', ' \\+ ', effect),
                effect=gsub('_', '\\\\_', effect),
                pLI=latexLinkPli(pLI, gene),
                gene=latexLinkGenes(gene)) %>%
    kable(booktabs=TRUE, escape=FALSE) %>% column_spec(2, width=col.width) %>% cat
})
```

# By gene

## Genes of interest

Genes of interest: `r sort(genes$hgnc_symbol)`.

```{r geneselpg, results='asis'}
high.df = eff.df %>% filter(impact=='HIGH', gene %in% genes$hgnc_symbol) %>%
  merge(pli.df, all.x=TRUE) %>% arrange(desc(pLI))
  
## Write a section for each variant
t = lapply(unique(high.df$gene), function(gene.name){
  df = high.df %>% filter(gene==gene.name)
  vars = tibble()
  if(any(df$method=='Sniffles')){
    id.ii = df %>% filter(method=='Sniffles') %>% .$id %>% unique
    vcf = vcf.sn[id.ii]
    vars = tibble(
      id=id.ii,
      pos=latexLinkPos(as.character(seqnames(rowRanges(vcf))), start(rowRanges(vcf))),
      type=info(vcf)$SVTYPE,
      size=info(vcf)$SVLEN,
      ## quality=rowRanges(vcf)$QUAL,
      reads=info(vcf)$RE,
      gnomad.freq=info(vcf)$AFMAX,
      method=info(vcf)$METHOD,
      nb.methods=info(vcf)$METHODS) %>% rbind(vars)
  }
  if(any(df$method=='SVIM')){
    id.ii = df %>% filter(method=='SVIM') %>% .$id %>% unique
    vcf = vcf.sv[id.ii]
    vars = tibble(
      id=id.ii,
      pos=latexLinkPos(as.character(seqnames(rowRanges(vcf))), start(rowRanges(vcf))),
      type=info(vcf)$SVTYPE,
      size=info(vcf)$SVLEN,
      ## quality=rowRanges(vcf)$QUAL,
      reads=info(vcf)$SUPPORT,
      gnomad.freq=info(vcf)$AFMAX,
      method=info(vcf)$METHOD,
      nb.methods=info(vcf)$METHODS) %>% rbind(vars)
  }
  vars = df %>% mutate(effect=gsub('\\+', ' \\+ ', effect),
                       effect=gsub('_', '\\\\_', effect)) %>%
    group_by(method, id) %>% summarize(effect=paste(effect, collapse='; ')) %>%
    merge(vars, .) %>% ungroup %>% select(-id)
  ## Write Markdown/Latex section
  cat('\n\n### ', mdLinkGenes(gene.name), '\n\n')
  cat('\n\n- pLI score: ', mdLinkPli(df$pLI[1], gene.name), '\n\n')
  eff.coln = which(colnames(vars)=='effect')
  vars %>% kable(booktabs=TRUE, escape=FALSE) %>%
    column_spec(eff.coln, width=col.width.gene) %>% cat
})
```

## Other protein-coding genes

```{r proteincodingpg, results='asis'}
high.df = eff.df %>% filter(impact=='HIGH', target.type=='protein_coding',
                            !(gene %in% genes$hgnc_symbol)) %>%
  merge(pli.df, all.x=TRUE) %>% arrange(desc(pLI))
  
## Write a section for each variant
t = lapply(unique(high.df$gene), function(gene.name){
  df = high.df %>% filter(gene==gene.name)
  vars = tibble()
  if(any(df$method=='Sniffles')){
    id.ii = df %>% filter(method=='Sniffles') %>% .$id %>% unique
    vcf = vcf.sn[id.ii]
    vars = tibble(
      id=id.ii,
      pos=latexLinkPos(as.character(seqnames(rowRanges(vcf))), start(rowRanges(vcf))),
      type=info(vcf)$SVTYPE,
      size=info(vcf)$SVLEN,
      ## quality=rowRanges(vcf)$QUAL,
      reads=info(vcf)$RE,
      gnomad.freq=info(vcf)$AFMAX,
      method=info(vcf)$METHOD,
      nb.methods=info(vcf)$METHODS) %>% rbind(vars)
  }
  if(any(df$method=='SVIM')){
    id.ii = df %>% filter(method=='SVIM') %>% .$id %>% unique
    vcf = vcf.sv[id.ii]
    vars = tibble(
      id=id.ii,
      pos=latexLinkPos(as.character(seqnames(rowRanges(vcf))), start(rowRanges(vcf))),
      type=info(vcf)$SVTYPE,
      size=info(vcf)$SVLEN,
      ## quality=rowRanges(vcf)$QUAL,
      reads=info(vcf)$SUPPORT,
      gnomad.freq=info(vcf)$AFMAX,
      method=info(vcf)$METHOD,
      nb.methods=info(vcf)$METHODS) %>% rbind(vars)
  }
  vars = df %>% mutate(effect=gsub('\\+', ' \\+ ', effect),
                       effect=gsub('_', '\\\\_', effect)) %>%
    group_by(method, id) %>% summarize(effect=paste(unique(effect), collapse='; ')) %>%
    merge(vars, .) %>% ungroup %>% select(-id)
  ## Write Markdown/Latex section
  cat('\n\n### ', mdLinkGenes(gene.name), '\n\n')
  cat('\n\n- pLI score: ', mdLinkPli(df$pLI[1], gene.name), '\n\n')
  eff.coln = which(colnames(vars)=='effect')
  vars %>% kable(booktabs=TRUE, escape=FALSE) %>%
    column_spec(eff.coln, width=col.width.gene) %>% cat
})
```
