---
title: Structural Variation Report
author: UPD-UCSC
header-includes:
- \usepackage{booktabs}
- \usepackage{makecell}
- \usepackage{makecell}
urlcolor: teal
output:
  pdf_document:
   keep_tex: true 
---

```{r include=FALSE}
knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE, fig.width=10)
```

```{r setup}
## Read input arguments
## 1: VCF file from Sniffles
args = commandArgs(TRUE)
args = c('na12878-chr11.sniffles.ann.freqGnomADcov10.vcf','na12878-chr11.svim.ann.freqGnomADcov10.vcf','gene_position_info.txt')

## Load packages
library(sveval)
library(VariantAnnotation)
library(dplyr)
library(knitr)
library(kableExtra)

## Parse the SnpEff field EFF
parseEff <- function(eff){
  effect = gsub('(.*)\\(.*\\)', '\\1', eff)
  target = gsub('.*\\((.*)\\)', '\\1', eff)
  target = strsplit(target, '\\|')[[1]]
  return(tibble(effect=effect, impact=target[[1]], gene=target[[6]],
              target.type=target[[7]]))
}
mdLinkGenes <- function(genes){
  paste0('[', genes, '](https://www.ncbi.nlm.nih.gov/gene?term=(', genes, '%5BGene%20Name%5D)%20AND%20Human%5BOrganism%5D)')
}
latexLinkGenes <- function(genes){
  paste0('{\\href{https://www.ncbi.nlm.nih.gov/gene?term=(', genes, '%5BGene%20Name%5D)%20AND%20Human%5BOrganism%5D}{', genes, '}}')
}
latexLinkPos <- function(chr, pos){
  paste0('{\\href{https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&position=', chr, '%3A', pos-500, '-', pos+500, '}{', chr, ':', pos, '}}')
}
col.width='13cm'

## Read both VCFs
vcf.sn = readSVvcf(args[1], vcf.object=TRUE)
vcf.sv = readSVvcf(args[2], vcf.object=TRUE)

## Make GRanges object for the overlap
gr.sn = rowRanges(vcf.sn)
end(gr.sn) = info(vcf.sn)$END
gr.sn$size = info(vcf.sn)$SIZE
gr.sn$type = info(vcf.sn)$SVTYPE
gr.sn$GT = 'hom'
gr.sn$ID = 1:length(vcf.sn)
gr.sv = rowRanges(vcf.sv)
end(gr.sv) = info(vcf.sv)$END
gr.sv$size = info(vcf.sv)$SIZE
gr.sv$type = info(vcf.sv)$SVTYPE
gr.sv$GT = 'hom'
gr.sv$ID = 1:length(vcf.sv)

## Overlap and add a column with overlap
ol.o = svOverlap(gr.sn, gr.sv)
info(vcf.sn)$METHODS = 1
info(vcf.sn)$METHODS[ol.o$query$ID] = 2
info(vcf.sv)$METHODS = 1
info(vcf.sv)$METHODS[ol.o$subject$ID] = 2
info(vcf.sn)$METHOD = 'Sniffles'
info(vcf.sv)$METHOD = 'SVIM'

## Read gene list
genes = read.table(args[3], as.is=TRUE, header=TRUE)

## DEBUG
genes$hgnc_symbol[1] = 'MUC5B'
```

```{r highimpact}
eff.df = lapply(list(vcf.sn, vcf.sv), function(vcf){
  high.ii = which(unlist(lapply(info(vcf)$EFF, function(effs) any(grepl('HIGH', effs)))))
  eff.df = lapply(high.ii, function(ii){
    effs = lapply(info(vcf)$EFF[[ii]], parseEff)
    effs = do.call(rbind, effs)
    effs$id = ii
    effs
  })
  eff.df = do.call(rbind, eff.df)
  eff.df$method = info(vcf)$METHOD[1]
  eff.df
})
eff.df = do.call(rbind, eff.df)
```

The structural variants found by [Sniffles](https://github.com/fritzsedlazeck/Sniffles) and [SVIM](https://github.com/eldariont/svim) were annotated by [SnpEff](http://snpeff.sourceforge.net/) and with the frequency of variants in the [gnoma-SV catalog](https://macarthurlab.org/2019/03/20/structural-variants-in-gnomad/).

The variants' "effects" in this report come from SnpEff and follow the format [described in SnpEff documentation](http://snpeff.sourceforge.net/SnpEff_manual.html#eff).

# High impact variants in genes of interest

Genes of interest: `r sort(genes$hgnc_symbol)`.

Note: for testing purposes I added *MUC5B* to the list of genes.

```{r genesel, results='asis'}
high.df = eff.df %>% filter(impact=='HIGH', gene %in% genes$hgnc_symbol) %>% arrange(gene)
high.ii = high.df %>% select(id, method) %>% unique

t = lapply(1:nrow(high.ii), function(ii){
  id.ii = high.ii$id[ii]
  if(high.ii$method[ii] == 'Sniffles'){
    vcf = vcf.sn[id.ii]
    reads = info(vcf.sn)$RE[id.ii]
  } else {
    vcf = vcf.sv[id.ii]
    reads = info(vcf.sv)$SUPPORT[id.ii]
  }
  res = tibble(
    pos=latexLinkPos(as.character(seqnames(rowRanges(vcf))), start(rowRanges(vcf))),
    type=info(vcf)$SVTYPE,
    size=info(vcf)$SVLEN,
    ## quality=rowRanges(vcf)$QUAL,
    reads=reads,
    gnomad.freq=info(vcf)$AFMAX,
    method=info(vcf)$METHOD,
    nb.methods=info(vcf)$METHODS)
  df = high.df %>% filter(id==id.ii) %>% select(gene, effect) %>% unique
  ## PDF version
  cat('\n\n## ', res$type, 'in', paste(mdLinkGenes(unique(df$gene)), collapse=' '), '\n\n')
  res %>% kable(booktabs=TRUE, escape=FALSE) %>% cat
  cat('\n\n### Effect(s) \n\n')
  df %>% mutate(effect=gsub('\\+', ' \\+ ', effect),
                effect=gsub('_', '\\\\_', effect),
                gene=latexLinkGenes(gene)) %>%
    kable(booktabs=TRUE, escape=FALSE) %>% column_spec(2, width=col.width) %>% cat
})
```

# High impact variants affecting other protein-coding genes

```{r proteincoding, results='asis'}
high.df = eff.df %>% filter(impact=='HIGH', target.type=='protein_coding',
                            !(gene %in% genes$hgnc_symbol)) %>% arrange(gene)
high.ii = high.df %>% select(id, method) %>% unique

t = lapply(1:nrow(high.ii), function(ii){
  id.ii = high.ii$id[ii]
  if(high.ii$method[ii] == 'Sniffles'){
    vcf = vcf.sn[id.ii]
    reads = info(vcf.sn)$RE[id.ii]
  } else {
    vcf = vcf.sv[id.ii]
    reads = info(vcf.sv)$SUPPORT[id.ii]
  }
  res = tibble(
    pos=latexLinkPos(as.character(seqnames(rowRanges(vcf))), start(rowRanges(vcf))),
    type=info(vcf)$SVTYPE,
    size=info(vcf)$SVLEN,
    ## quality=rowRanges(vcf)$QUAL,
    reads=reads,
    gnomad.freq=info(vcf)$AFMAX,
    method=info(vcf)$METHOD,
    nb.methods=info(vcf)$METHODS)
  df = high.df %>% filter(id==id.ii) %>% select(gene, effect) %>% unique
  ## PDF version
  cat('\n\n## ', res$type, 'in', paste(mdLinkGenes(unique(df$gene)), collapse=' '), '\n\n')
  res %>% kable(booktabs=TRUE, escape=FALSE) %>% cat
  cat('\n\n### Effect(s) \n\n')
  df %>% mutate(effect=gsub('\\+', ' \\+ ', effect),
                effect=gsub('_', '\\\\_', effect),
                gene=latexLinkGenes(gene)) %>%
    kable(booktabs=TRUE, escape=FALSE) %>% column_spec(2, width=col.width) %>% cat
})
```

# Next

- Add LoF intolerance score from gnomAD.
